src/layers/python.rs

```diff
@@ -0,0 +1,582 @@
+use crate::packaging_tool_versions::PackagingToolVersions;
+use crate::python_version::PythonVersion;
+use crate::utils::{self, CommandError, DownloadUnpackArchiveError};
+use crate::{BuildpackError, PythonBuildpack};
+use libcnb::build::BuildContext;
+use libcnb::data::buildpack::StackId;
+use libcnb::data::layer_content_metadata::LayerTypes;
+use libcnb::layer::{ExistingLayerStrategy, Layer, LayerData, LayerResult, LayerResultBuilder};
+use libcnb::layer_env::{LayerEnv, ModificationBehavior, Scope};
+use libcnb::{Buildpack, Env};
+use libherokubuildpack::log::log_info;
+use serde::{Deserialize, Serialize};
+use std::fs::Permissions;
+use std::os::unix::prelude::PermissionsExt;
+use std::path::{Path, PathBuf};
+use std::process::Command;
+use std::{fs, io};
+
+/// Layer containing the Python runtime, and the packages `pip`, `setuptools` and `wheel`.
+///
+/// We install both Python and the packaging tools into the same layer, since:
+///  - We don't want to mix buildpack/packaging dependencies with the app's own dependencies
+///    (for a start, we need pip installed to even install the user's own dependencies, plus
+///    want to keep caching separate), so cannot install the packaging tools into the user
+///    site-packages directory.
+///  - We don't want to install the packaging tools into an arbitrary directory added to
+///    `PYTHONPATH`, since directories added to `PYTHONPATH` take precedence over the Python
+///    stdlib (unlike the system or user site-packages directories), and so can result in hard
+///    to debug stdlib shadowing problems that users won't encounter locally.
+///  - This leaves just the system site-packages directory, which exists within the Python
+///    installation directory and Python does not support moving it elsewhere.
+///  - It matches what both local and official Docker image environments do.
+pub(crate) struct PythonLayer<'a> {
```

Comment:
It's a bit annoying working with structs that don't own their data. Cloning is minimal overhead in this context and I'm mindful that people who primarily dev in python (and not in Rust) may try to submit a PR and that might be a stumbling block. (For an analogy: Perhaps you remember how annoying working on libcnb internals with the generic in `libcnb::Error<E>` was when first getting started). 

I think that using a reference here might be the more "rust-y" way to do things, (clone is bad, burn with fire 🔥). But in our scripting context, I think (some) dev speed takes precedence. Shorter: I don't think the gain of no clone is worth having to put up with lifetime annotations for the rest of the file.

Reply:
Also Debug all structs by default please.

Reply:
Bikeshed should be blue! It's not a blocker. Feel free to resolve.

It's a bit of an open question how similar should buildpack internal/style be? Should we update all buildpack layers use non-owned data?

Reply:
For some things there is a definite advantage in standardising, particularly when there is either a clear best practice (eg: "use rustfmt on our codebases") or the decision matters from a product point of view (eg: consistent log output styles between buildpacks).

However, I think there are always going to be things that fall outside of the scope of linters/formatters and just come down to whatever implementation details happen to have been chosen - e.g. use of an if statement vs a match statement, or in this case whether a struct owns its data or not. For these I'm not sure I see the benefit on standardising - and also maybe there are more important things to discuss instead of them, during review?